### PR TITLE
diary, groups, heap: version negotiation

### LINKED ISSUE #950: diary: write version negotiation logic

### PR DESCRIPTION
**Pull request by ryjm**
Includes state update for `%net` migration.

Resolves #950, #578, and #579 (WIP)

### CODE REVIEW

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:11]: should cast with ` syntax

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:89]: missing state renegotiation on upgrade

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:213]: resolved OOB, subs are definitely removed before kick

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:218]: note that not crashing here is correct, because crash paths in %fact handling can often lead to resubscribe loops and in the general will buy you more flexibility when you need to change the semantics.

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:226]: please check that `=(src.bowl p.flag)`

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:525]: rewrite with just `?.`

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:531]: missing giving an %epic as the first fact on every networked diary subscription

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:13]: same comment about tec

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:43]: just don't

File: desk/app/diary.hoon
Comment (liam-fitzgerald) [inline desk/app/diary.hoon:33]: should be `!>([okay state])`

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:22]: remove from state and into the devasing

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:184]: `=+  !<([=old=epic:e old=versioned-state] vase)`

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:250]: same line pls

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:310]: same as %diary comment

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:323]: same commen as %diary

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:693]: same w/ `?.`

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:692]: actually, you can't do this my bad.

The correct fix here is to remove the %load case in `$net` which will make this assertion correct, and just add a flag in the `%sub` case, to determine whether or not it has been loaded

Comment (liam-fitzgerald) [reply desk/app/groups.hoon:692]: this will also need to be propagated to the other agents, inc. chat (mea culpa)

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:694]: should be `(go-sub load.net)` w/ above changes w.r.t removal of %load

File: desk/app/groups.hoon
Comment (liam-fitzgerald) [inline desk/app/groups.hoon:785]:
```hoon
=.  net
  ?:   ?=(%pub -.net)
     pub/(put:log-on:g p.net time diff)
  [%sub time %chi ~]
```

File: desk/sur/diary-1.hoon
Comment (liam-fitzgerald) [inline desk/sur/diary-1.hoon:8]: drop this file?

File: desk/sur/groups.hoon
Comment (liam-fitzgerald) [inline desk/sur/groups.hoon:284]: aforementioned removal of %load discussed above

**Review by liam-fitzgerald — CHANGES_REQUESTED** [review]
please address comments

**Review by arthyn — CHANGES_REQUESTED** [review]
So we definitely need to remove the %load case from chat, and then are we nuking binnec again instead of migrating those changes? If we are then this is good to go after fixing chat types. Also was %load the reason the watch-acks failed? I don't know if we ever explicitly stated anywhere.

File: desk/sur/chat.hoon
Comment (ryjm) [reply desk/sur/chat.hoon:1]: > So we definitely need to remove the %load case from chat, and then are we nuking binnec again instead of migrating those changes? If we are then this is good to go after fixing chat types. Also was %load the reason the watch-acks failed? I don't know if we ever explicitly stated anywhere.

 the watch acks weren't harmful actually, was just that swapped line. i was able to trigger the watch ack failure for the case that i was worried about and i'm confident they aren't harmful. would be nice to get rid of them though.

%load is removed now, definitely need a thorough review on the chat changes  though.

File: desk/app/groups.hoon
Comment (ryjm) [reply desk/app/groups.hoon:662]: done, with some lingering questions. no way to migrate from old %load subscriptions since we don't have a @p to use - unless we can grab it from somewhere else in the state.

File: desk/sur/chat.hoon
Comment (arthyn) [reply desk/sur/chat.hoon:1]: I'm not seeing load removed. where did you mean?

File: desk/app/diary.hoon
Comment (ryjm) [reply desk/app/diary.hoon:88]: nukes state 😈

it will only do this if the old state does not correspond to any version we know about - i.e if you update `state-0` yourself. we wouldn't be able to migrate that state anyway. we could crash here too, but then you'd have to either nuke manually or write a custom migration to get it running. also with #1065 you'll never fully lose state anyway.

File: desk/sur/chat.hoon
Comment (ryjm) [reply desk/sur/chat.hoon:1]: https://github.com/tloncorp/landscape-apps/pull/1031/files/f761a033e195e2d341528659c046c2052c41ebbb#diff-9c76c0009e3d5ee93a5aefccb8c8c6777723e4c6c99dabb5b24891fae8466d0aR232

Comment (arthyn) [reply desk/sur/chat.hoon:1]: I meant in chat but I saw it in your other PR so I think they just need to get merged together

File: desk/app/chat.hoon
Comment (ryjm) [inline desk/app/chat.hoon:439]: @liam-fitzgerald we had these conditionals wrong right? just checking

File: desk/app/chat.hoon
Comment (arthyn) [inline desk/app/chat.hoon:175]: I don't think this is possible now right? since it was removed

**Review by arthyn — COMMENTED** [review]
just the one question/change other than that I think this is all correct, but I'd like to be walked through it before we merge

File: desk/app/heap.hoon
Comment (midsum-salrux) [inline desk/app/heap.hoon:245]: Is it possible for the publisher to skip directly from behind the subscriber to ahead of the subscriber? If so, will this be handled correctly? Say the subscriber is at 5, and the publisher goes from 4 to 6 without us getting a `fact` for 5.

Comment (liam-fitzgerald) [reply desk/app/heap.hoon:245]: The fix for this is announcing the epic on watch paths, before giving the %chat-logs or w.e.

**Comment by midsum-salrux** [comment]
Doing some testing and I noticed the `saga` get out of sync as you change the `epic` of the publisher and subscriber. eg if the pub goes to 1 while the sub remains at zero, it will stay at [%dex ver=1] and not update

for example if the pub goes to 1 while the sub remains at zero, it will stay at [%dex ver=1] and not update

I got this error message as that was happening:

```
>=
%chat: on-agent on wire /epic, %fact
out=[acked=%.y path=/chat/~lapper-dozzod-tomdys/version-negotiation/updates]
subscribe wire not unique
%chat
/chat/~lapper-dozzod-tomdys/version-negotiation/updates
[p=~lapper-dozzod-tomdys q=%chat]
%chat: on-fail with term %watch-not-unique
error in %chat
%watch-not-unique
! subscribe wire not unique
! %chat
! /chat/~lapper-dozzod-tomdys/version-negotiation/updates
! [p=~lapper-dozzod-tomdys q=%chat]
%chat: on-agent on wire /chat/~lapper-dozzod-tomdys/version-negotiation/updates, %fact
%chat: on-poke with mark %dbug
```

**Comment by arthyn** [comment]
@jamesacklin @ryjm @liam-fitzgerald alright I merged secret groups into this branch and cleaned up/reset the state migration since we're nuking. The only thing we need is to make sure the subscription mechanism works.

**Comment by midsum-salrux** [comment]
Looks like I'm still getting a kick loop when the epics are out of sync. 
```
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %watch-ack
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %kick
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %watch-ack
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %kick
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %watch-ack
%groups: on-agent on wire /gangs/~lapper-dozzod-tomdys/verneg/preview, %kick
```

Subscriber `+dbug`
```
[ [ %0
      groups
    { [ p=[p=~lapper-dozzod-tomdys q=%verneg]
          q
        [ [%sub p=~2022.11.21..15.22.58..fb23 load=%.n saga=[%dex ver=2]]
          fleet={}
          cabals={}
          zones={}
          zone-ord=~
          bloc={}
          channels={}
          cordon=[%open ban=[ships={} ranks={}]]
          secret=%.y
          meta=[title='' description='' image='' cover='']
        ]
      ]
    }
      xeno
    { [ p=[p=~lapper-dozzod-tomdys q=%verneg]
          q
        [ cam=[~ [join-all=%.y progress=%watching]]
          pev=~
          vit=[~ [p=[p=~lapper-dozzod-tomdys q=%verneg] q=~locdev-midsum-salrux]]
        ]
      ]
    }
  ]
  0
]
```

**Comment by arthyn** [comment]
@midsum-salrux may want to checkout out @liam-fitzgerald 's pr here: https://github.com/tloncorp/landscape-apps/pull/1210

**Review by arthyn — APPROVED** [review]

**Pull request merged**
